tests/UnifiedSpecTests/Constraint/Match.php

```diff
@@ -0,0 +1,470 @@
+        $this->ignoreExtraKeysInEmbedded = $ignoreExtraKeysInEmbedded;
+        $this->comparatorFactory = Factory::getInstance();
+    }
+
+    public function evaluate($other, $description = '', $returnResult = false)
+    {
+        $other = self::prepare($other, true);
+        $success = false;
+        $this->lastFailure = null;
+
+        try {
+            $this->assertMatches($this->value, $other, $this->ignoreExtraKeysInRoot);
+            $success = true;
+        } catch (RuntimeException $e) {
+            $this->lastFailure = new ComparisonFailure(
+                $this->value,
+                $other,
+                $this->exporter()->export($this->value),
+                $this->exporter()->export($other),
+                false,
+                $e->getMessage()
+            );
+        }
+
+        if ($returnResult) {
+            return $success;
+        }
+
+        if (! $success) {
+            $this->fail($other, $description, $this->lastFailure);
+        }
+    }
+
+    private function assertEquals($expected, $actual, string $keyPath)
+    {
+        $expectedType = is_object($expected) ? get_class($expected) : gettype($expected);
+        $actualType = is_object($actual) ? get_class($actual) : gettype($actual);
+
+        // Workaround for ObjectComparator printing the whole actual object
+        if ($expectedType !== $actualType) {
+            throw new ComparisonFailure(
+                $expected,
+                $actual,
+                '',
+                '',
+                false,
+                sprintf(
+                    '%s%s is not instance of expected type "%s".',
+                    empty($keyPath) ? '' : sprintf('Field path "%s": ', $keyPath),
+                    $this->exporter()->shortenedExport($actual),
+                    $expectedType
+                )
+            );
+        }
+
+        try {
+            $this->comparatorFactory->getComparatorFor($expected, $actual)->assertEquals($expected, $actual);
+        } catch (ComparisonFailure $failure) {
+            throw new ComparisonFailure(
+                $expected,
+                $actual,
+                '',
+                '',
+                false,
+                (empty($keyPath) ? '' : sprintf('Field path "%s": ', $keyPath)) . $failure->getMessage()
+            );
+        }
+    }
+
+    /**
+     * Compares two BSON values recursively.
+     *
+     * @param mixed  $expected
+     * @param mixed  $actual
+     * @param bool   $ignoreExtraKeys
+     * @param string $keyPath
+     * @throws RuntimeException if the documents do not match
+     */
+    private function assertMatches($expected, $actual, bool $ignoreExtraKeys, $keyPath = '')
+    {
+        if (! $expected instanceof BSONDocument && ! $expected instanceof BSONArray) {
+            $this->assertEquals($expected, $actual, $keyPath);
+
+            return;
+        }
+
+        if ($expected instanceof BSONArray) {
```

Comment:
Maybe we can extract some of this logic to their own methods or even own constraint classes (if it wouldn't lead to too much code duplication)? As is, the method is a little hard to follow since it covers so many different use-cases.

Reply:
Aye, I did intend on breaking this up into separate methods for the array/object cases.

I think there may be diminishing returns on turning everything into a constraint, since our only use case is calling `matches()` to get a boolean result. Ideally, we could rely on the constraint's own semantic error and concatenate our field path info from MatchConstraint, but I didn't want to waste more time trying to figure out how PHPUnit's Constraint API works (and perhaps this is made more difficult due to the BC brdige we're using).

Reply:
Just pushed fixes and tests for both constraint classes, so PTAL.